iothub/device/samples/how to guides/DeviceReconnectionSample/DeviceReconnectionSample.cs

```diff
@@ -115,40 +121,29 @@ private async Task InitializeAndSetupClientAsync(CancellationToken cancellationT
                             {
                                 await s_deviceClient.CloseAsync(cancellationToken);
                             }
-                            catch (UnauthorizedException) { } // if the previous token is now invalid, this call may fail
+                            catch (UnauthorizedException) { } // If the previous token is now invalid, this call may fail
                             s_deviceClient.Dispose();
                         }
 
-                        s_deviceClient = DeviceClient.CreateFromConnectionString(_deviceConnectionStrings.First(), _transportType, _clientOptions);
+                        s_deviceClient = DeviceClient.CreateFromConnectionString(_deviceConnectionStrings.First(), _transportType, s_clientOptions);
                         s_deviceClient.SetConnectionStatusChangesHandler(ConnectionStatusChangeHandlerAsync);
+                        s_deviceClient.SetRetryPolicy(_customRetryPolicy);
                         _logger.LogDebug("Initialized the client instance.");
+
+                        // Force connection now.
+                        // We have set the "shouldExecuteOperation" function to always try to open the connection.
+                        // OpenAsync() is an idempotent call, it has the same effect if called once or multiple times on the same client.
+                        await s_deviceClient.OpenAsync(cancellationToken);
```

Comment:
Open should be inside the semaphore, because whether we're connected or not determines if we go in. So, another thread waiting on this semaphore should not enter after we've initialized and opened, thus changing the state.